carla/models/api/mlmodel.py

```diff
@@ -82,3 +82,65 @@ def predict_proba(self, x):
             Ml model prediction with shape N x 2
         """
         pass
+
+    @property
+    @abstractmethod
+    def scaler(self):
+        """
+        Yields a fitted normalizer.
+        To keep a consistent and correct normalization for the ML model
+
+        We recommend to use sklear normalization preprocessing
+
+        Returns
+        -------
+        Fitted scaler for normalization
+        """
+        pass
+
+    @abstractmethod
+    def set_scaler(self, data):
```

Comment:
I am unsure this really should be a public API method. I think that should be abstracted away from the user. We could think of a design e.g.

```
class MLModel(self, data):

def __init__(): 
  self._scaler = None
  self.data = data

@property
def scaler(self):
    return self._scaler

@property
def data(self):
    return self._data
    
@data.setter
def data(self, data):
    self._data = data
    self._scaler = preprocessing.MinMaxScaler().fit(data.raw[self._continuous])
   

```
That would mean the user wouldn't have to bother setting scalers themselfes but still ensuring the scaler is always up to date in case we'd change the data

Reply:
Without a setter for the scaler, our ML api would not be usable for users with models, which are already trained with a different scaler. We'd always enforce a MinMax scaler, wouldn't we?

Reply:
Good point, didn't think of that!

To make it more versatile and allow users to actually set their own scaler we could fit the "default" MinMaxScaler (can be externded later to other scalers) in the constructor. Then we can change the setter to accept a scaler and "override" the default scaler? Something like this:

```
class MLModel(self, data):

def __init__(..., scaling_method="MinMax"): 
  self.data = data
  
  if scaling_method == "MinMax":
      fitted_scaler = preprocessing.MinMaxScaler().fit(data.raw[self._continuous])
      self.scaler(fitted_scaler)

@property
def data(self):
    return self._data
    
@data.setter
def data(self, data):
    self._data = data

@property
def scaler(self):
    return self._scaler

@scaler.setter
def scaler(self, scaler):
    self._scaler = scaler
```

Then the user could either just do

```
catalog_model = MLModel(data,...)

my_custom_scaled_model = MlModel(data,...)
my_custom_scaled_model.scaler = my_fitted_scaler
```

Would that work?

We could even extend it to allow to pass the scaler/encoder to be passed via the contstructor aswell.

Reply:
This sounds great 👍 
Further, I thought about enforcing to have only sklearn scaler and encoder. So we could rely on built-in functions like transform and inverse_transform. Sklearn should provide enough different methods and has a good support.
WDYT?

Reply:
I think that is a good solution. Sklearn has the option for people to implement their own custom scaler/encoder.

Reply:
Really, this makes everything even better 😄

Reply:
Yep https://scikit-learn.org/stable/modules/preprocessing.html#custom-transformers

Reply:
Lets do it!